src/rabbit_delayed_message_kv_store.erl

```diff
@@ -0,0 +1,242 @@
+%% This Source Code Form is subject to the terms of the Mozilla Public
+%% License, v. 2.0. If a copy of the MPL was not distributed with this
+%% file, You can obtain one at https://mozilla.org/MPL/2.0/.
+%%
+%%  Copyright (c) 2007-2020 VMware, Inc. or its affiliates.  All rights reserved.
+%%
+
+%% NOTE that this module uses os:timestamp/0 but in the future Erlang
+%% will have a new time API.
+%% See:
+%% https://www.erlang.org/documentation/doc-7.0-rc1/erts-7.0/doc/html/erlang.html#now-0
+%% and
+%% https://www.erlang.org/documentation/doc-7.0-rc1/erts-7.0/doc/html/time_correction.html
+
+-module(rabbit_delayed_message_kv_store).
+-include_lib("rabbit_common/include/rabbit.hrl").
+-include_lib("kernel/include/logger.hrl").
+-include_lib("rabbit_common/include/logging.hrl").
+
+-behaviour(gen_server).
+
+-export([start_link/0, leveled_bookie_start_link/0]).
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2,
+         code_change/3]).
+
+
+-export([setup/0, maybe_make_cluster/0, maybe_resize_cluster/0]).
+
+-export([
+         write/2,
+         delete/1,
+         do_write/2,
+         do_delete/1,
+         do_take/1
+         ]).
+
+-record(state, {kv_store_pid
+               }).
+
+-define(RA_SYSTEM, delayed_message_exchange).
+-define(RA_CLUSTER_NAME, dmx_kv_cluster).
+
+write(Key, Value) ->
+    rabbit_delayed_message_machine:write({?RA_CLUSTER_NAME, node()},
+                                         Key,
+                                         Value).
+
+delete(Key) ->
+    rabbit_delayed_message_machine:delete({?RA_CLUSTER_NAME, node()},
+                                          Key).
+
+
+setup() ->
+    ok = ensure_ra_system_started(?RA_SYSTEM).
+
+maybe_make_cluster() ->
+    Local = {?RA_CLUSTER_NAME, node()},
+    Nodes = rabbit_nodes:list_reachable(),
+    %% This might be overkill?
+    case whereis(?RA_CLUSTER_NAME) of
+        undefined ->
+            global:set_lock({?RA_CLUSTER_NAME, self()}),
+            case ra:restart_server(?RA_SYSTEM, Local) of
+                {error, Reason} when Reason == not_started orelse
+                                     Reason == name_not_registered ->
+                    OtherNodes = Nodes -- [node()],
+                    rabbit_log:debug(">>>> OTHER NODES ~p", [OtherNodes]),
+                    case lists:filter(
+                           fun(N) ->
+                                   erpc:call(N, erlang, whereis, [?RA_CLUSTER_NAME]) =/= undefined
+                           end, OtherNodes) of
+                        [] ->
+                            ra:start_cluster(?RA_SYSTEM, [make_ra_conf(Node, Nodes) || Node <-  Nodes]);
+                        _ ->
+                            ok
+                    end;
+                _ ->
+                    ok
+            end,
+            global:del_lock({?RA_CLUSTER_NAME, self()});
+        _ ->
+            maybe_resize_cluster(),
+            ok
+    end,
+    ok.
+
+
+maybe_resize_cluster() ->
```

Comment:
in case of stream_coordinator the leader is moved from a node which is put in maintenance, maybe that behaviour should also be replicated here